synapse/handlers/e2e_keys.py

```diff
@@ -581,6 +581,12 @@ async def claim_local_one_time_keys(
             An iterable of maps of user ID -> a map device ID -> a map of key ID -> JSON bytes.
         """
 
+        # Cap the number of OTKs that can be claimed at once to avoid abuse.
+        local_query = [
+            (user_id, device_id, algorithm, min(count, 5))
```

Comment:
I am trying to understand how it is fine to do so regarding client behavior.

I think we are ok because current spec says:
> If the homeserver could be reached, but the user or device was unknown, no failure is recorded. Instead, the corresponding user or device is missing from the one_time_keys result.

Which can be extended to the `not enough keys to match count` use case: clients are expected to handle missing keys in the result without relying on explicit errors.

Reply:
> Which can be extended to the `not enough keys to match count` use case: clients are expected to handle missing keys in the result without relying on explicit errors.

Yeah, that was kind of my thought -- if you got back fewer keys than expected then it is up to you (as the client) to deal with that.

Note that this now also always includes the fallback, so you can still do something if you have fewer keys.

Reply:
(I'll leave a comment on the MSC about this.)